core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnStringJoin.java

```diff
@@ -0,0 +1,139 @@
+          .name("arg1")
+          .type(IAnyAtomicItem.type())
+          .oneOrMore()
+          .build())
+      .returnType(IStringItem.type())
+      .returnOne()
+      .functionHandler(FnStringJoin::execute)
+      .build();
+
+  static final IFunction SIGNATURE_TWO_ARG = IFunction.builder()
+      .name(NAME)
+      .namespace(MetapathConstants.NS_METAPATH_FUNCTIONS)
+      .deterministic()
+      .contextIndependent()
+      .focusIndependent()
+      .argument(IArgument.builder()
+          .name("arg1")
+          .type(IAnyAtomicItem.type())
+          .zeroOrMore()
+          .build())
+      .argument(IArgument.builder()
+          .name("arg2")
+          .type(IStringItem.type())
+          .zeroOrOne()
+          .build())
+      .returnType(IStringItem.type())
+      .returnOne()
+      .functionHandler(FnStringJoin::execute)
+      .build();
+
+  private FnStringJoin() {
+    // disable construction
+  }
+
+  @SuppressWarnings("unused")
+  @NonNull
+  private static ISequence<IStringItem> execute(
+      @NonNull IFunction function,
+      @NonNull List<ISequence<?>> arguments,
+      @NonNull DynamicContext dynamicContext,
+      IItem focus) {
+
+    ISequence<?> arg1 = arguments.get(0);
+    IStringItem arg2 = arguments.size() == 1 ? IStringItem.valueOf("")
+        : FunctionUtils.asTypeOrNull(arguments.get(1).getFirstItem(true));
+    String separator = arg2 == null ? "" : arg2.asString();
+
+    if (arg1.isEmpty()) {
+      return ISequence.of(IStringItem.valueOf(""));
+    }
+
+    return ISequence.of(stringJoin(ObjectUtils.notNull(arg1.stream()
+        .map(arg -> {
+          assert arg != null;
+          return (IAnyAtomicItem) arg;
+        })), separator));
+  }
+
+  /**
+   * An implementation of XPath 3.1 <a href=
+   * "https://www.w3.org/TR/xpath-functions-31/#func-string-join">fn:string-join</a>.
+   *
+   * @param items
+   *          the items to concatenate
+   * @return the atomized result
+   */
+  @NonNull
+  public static IStringItem stringJoinstringJoin(IAnyAtomicItem items, String separator) {
+    return stringJoin(ObjectUtils.notNull(Arrays.asList(items)), separator);
+  }
+
+  /**
+   * An implementation of XPath 3.1 <a href=
+   * "https://www.w3.org/TR/xpath-functions-31/#func-string-join">fn:string-join</a>.
+   *
+   * @param items
+   *          the items to concatenate
+   * @return the atomized result
+   */
+  @NonNull
+  public static IStringItem stringJoin(@NonNull List<? extends IAnyAtomicItem> items, String separator) {
+    return stringJoin(ObjectUtils.notNull(items.stream()), separator);
+  }
+
+  /**
+   * An implementation of XPath 3.1 <a href=
+   * "https://www.w3.org/TR/xpath-functions-31/#func-string-join">fn:string-join</a>.
+   *
+   * @param items
+   *          the items to concatenate
+   * @return the atomized result
+   */
+  @NonNull
+  public static IStringItem stringJoin(@NonNull Stream<? extends IAnyAtomicItem> items, String separator) {
```

Comment:
Should the stream "version" be private or should this be moved into the collection version? The collection version doesn't seem to do much other than provide a stream to this method.

Reply:
Also, maybe it would be good to just return a Java `String` here.

Reply:
It depends, will there be any value in making it an invokable utility function we can use later?

Reply:
Adjusted it